text/0099-otlp-http.md

```diff
@@ -0,0 +1,146 @@
+MAY gzip-encode the response and set "Content-Encoding: gzip" response header.
+
+The server SHOULD respond with success no sooner than after successfully
+decoding and validating the request.
+
+#### Failures
+
+If the processing of the request fails because the request contains data that
+cannot be decoded or is otherwise invalid and such failure is permanent then the
+server MUST respond with `HTTP 400 Bad Request`. The client MUST NOT retry the
+request when it receives `HTTP 400 Bad Request` response.
+
+#### Throttling
+
+If the server receives more requests than the client is allowed or the server is
+overloaded the server SHOULD respond with `HTTP 429 Too Many Requests` or
+`HTTP 503 Service Unavailable` and MAY include "Retry-After" header with a
+recommended time interval in seconds to wait before retrying. The client SHOULD
+honour the waiting interval specified in "Retry-After" header if it is present.
+If the client receives `HTTP 429` or `HTTP 503` response and "Retry-After"
+header is not present in the response then the client SHOULD implement an
+exponential backoff strategy between retries.
+
+#### All Other Responses
+
+All other HTTP responses that are not explicitly listed in this document should
+be treated according to HTTP specification.
+
+If the server disconnects without returning a response the client SHOULD retry
+and send the same request. The client SHOULD implement an exponential backoff
+strategy between retries to avoid overwhelming the server.
+
+### Connection
+
+If the client is unable to connect to the server the client SHOULD retry the
+connection using exponential backoff strategy between retries. The interval
+between retries must have a random jitter.
+
+The client SHOULD keep the connection alive between requests. The client SHOULD
+use "Connection: Keep-Alive" request header.
+
+Server implementations MAY handle OTLP/gRPC and OTLP/HTTP requests on the same
+port and multiplex the connections to the corresponding transport handler based
+on "Content-Type" request header.
+
+### Parallel Connections
+
+To achieve higher total throughput the client MAY send requests using several
+parallel HTTP connections. In that case the maximum number of parallel
+connections SHOULD be configurable.
+
+## Prior Art and Alternatives
+
+I have also considered HTTP/1.1+WebSocket transport. Experimental implementation
+of OTLP over WebSocket transport has shown that it typically has better
+performance than plain HTTP transport implementation (WebSocket uses less CPU,
+higher throughput in high latency connections). However WebSocket transport
+requires slightly more complicated implementation and WebSocket libraries are
+less ubiquitous than plain HTTP, which may make implementation in certain
+languages difficult or impossible.
+
+HTTP/1.1+WebSocket transport may be considered as a future transport for
+high-performance use cases as it exhibits better performance than OTLP/gRPC and
```

Comment:
Can you share any numbers how much better the performance with WebSocket is?

Reply:
Also I think we can share some numbers in this comment, but I would suggest that a written version in specs or collector protocol to be posted after we verify that we are not doing some trivial mistakes.

Reply:
Not to distract from this PR, just to satisfy curiosity :-) below is plain HTTP (this PR) vs WebSocket. Both using single HTTP/1.1 connection.

These are results from quick code I put together (which likely has bugs). No profiling or optimizations done, so should not be considered definitive (code here https://github.com/tigrannajaryan/exp-otelproto). 

WebSocket wins for small batches (in CPU) or high latencies (in throughput) are pretty big.

The only outlier/counterexample is very large batches (unclear why it loses in that case, I did not dig into it - can be done if we seriously consider WebSocket in the future).

```
64000 nano batches, 1 spans per batch, 10 attrs per span
OTLP/HTTP1.1/1                 64000 spans| 13.1 cpusec|  8.9 wallsec| 4885.5 batches/cpusec|  7165.3 batches/wallsec|204.7 cpuμs/span
WebSocket/Stream/Async/1       64000 spans|  3.2 cpusec|  1.4 wallsec|19937.7 batches/cpusec| 44880.6 batches/wallsec| 50.2 cpuμs/span

16000 tiny batches, 10 spans per batch, 10 attrs per span
OTLP/HTTP1.1/1                160000 spans|  7.4 cpusec|  4.9 wallsec| 2176.9 batches/cpusec|  3256.3 batches/wallsec| 45.9 cpuμs/span
WebSocket/Stream/Async/1      160000 spans|  2.8 cpusec|  1.5 wallsec| 5714.3 batches/cpusec| 10597.7 batches/wallsec| 17.5 cpuμs/span

8000 small batches, 100 spans per batch, 10 attrs per span
OTLP/HTTP1.1/1                800000 spans| 14.5 cpusec| 11.2 wallsec|  550.6 batches/cpusec|   714.0 batches/wallsec| 18.2 cpuμs/span
WebSocket/Stream/Async/1      800000 spans| 10.3 cpusec|  6.1 wallsec|  779.7 batches/cpusec|  1313.7 batches/wallsec| 12.8 cpuμs/span

800 large batches, 500 spans per batch, 10 attrs per span
OTLP/HTTP1.1/1                400000 spans|  5.2 cpusec|  4.5 wallsec|  154.4 batches/cpusec|   177.7 batches/wallsec| 12.9 cpuμs/span
WebSocket/Stream/Async/1      400000 spans|  4.7 cpusec|  2.9 wallsec|  169.9 batches/cpusec|   277.5 batches/wallsec| 11.8 cpuμs/span

100 very large batches, 5000 spans per batch, 10 attrs per span
OTLP/HTTP1.1/1                500000 spans|  5.5 cpusec|  5.3 wallsec|   18.1 batches/cpusec|    18.9 batches/wallsec| 11.1 cpuμs/span
WebSocket/Stream/Async/1      500000 spans|  8.0 cpusec|  3.6 wallsec|   12.5 batches/cpusec|    27.4 batches/wallsec| 16.0 cpuμs/span

2ms network roundtrip latency
800 large batches, 500 spans per batch, 10 attrs per span
OTLP/HTTP1.1/1                400000 spans|  5.1 cpusec|  6.3 wallsec|  155.3 batches/cpusec|   126.8 batches/wallsec| 12.9 cpuμs/span
WebSocket/Stream/Async/1      400000 spans|  4.8 cpusec|  3.0 wallsec|  167.7 batches/cpusec|   270.6 batches/wallsec| 11.9 cpuμs/span

20ms network roundtrip latency
400 large batches, 500 spans per batch, 10 attrs per span
OTLP/HTTP1.1/1                200000 spans|  2.6 cpusec| 10.8 wallsec|  151.5 batches/cpusec|    36.9 batches/wallsec| 13.2 cpuμs/span
WebSocket/Stream/Async/1      200000 spans|  2.3 cpusec|  1.6 wallsec|  170.9 batches/cpusec|   250.2 batches/wallsec| 11.7 cpuμs/span

200ms network roundtrip latency
40 large batches, 500 spans per batch, 10 attrs per span
OTLP/HTTP1.1/1                 20000 spans|  0.3 cpusec| 12.7 wallsec|  148.1 batches/cpusec|     3.2 batches/wallsec| 13.5 cpuμs/span
WebSocket/Stream/Async/1       20000 spans|  0.2 cpusec|  2.5 wallsec|  166.7 batches/cpusec|    15.9 batches/wallsec| 12.0 cpuμs/span

200ms network roundtrip latency
400 large batches, 500 spans per batch, 10 attrs per span
OTLP/HTTP1.1/1                200000 spans|  2.8 cpusec| 86.4 wallsec|  144.9 batches/cpusec|     4.6 batches/wallsec| 13.8 cpuμs/span
WebSocket/Stream/Async/1      200000 spans|  2.4 cpusec| 28.4 wallsec|  167.4 batches/cpusec|    14.1 batches/wallsec| 11.9 cpuμs/span
```